libsolidity/codegen/YulUtilFunctions.cpp

```diff
@@ -101,21 +102,35 @@ std::string YulUtilFunctions::copyToMemoryFunction(bool _fromCalldata, bool _cle
 			("cleanup", _cleanup)
 			.render();
 		}
+		// memory -> memory
 		else
 		{
-			return Whiskers(R"(
-				function <functionName>(src, dst, length) {
-					let i := 0
-					for { } lt(i, length) { i := add(i, 32) }
-					{
-						mstore(add(dst, i), mload(add(src, i)))
+			// use MCOPY (cancun+)
+			if (m_evmVersion.hasMcopy())
+				return Whiskers(R"(
+					function <functionName>(src, dst, length) {
+						mcopy(dst, src, length)
+						<?cleanup>mstore(add(dst, length), 0)</cleanup>
 					}
-					<?cleanup>mstore(add(dst, length), 0)</cleanup>
-				}
-			)")
-			("functionName", functionName)
-			("cleanup", _cleanup)
-			.render();
+				)")
+				("functionName", functionName)
+				("cleanup", _cleanup)
+				.render();
+			// mstore mload loop
+			else
+				return Whiskers(R"(
+					function <functionName>(src, dst, length) {
```

Comment:
This now occurs three times over... and each time with identical name and arguments... we could use the other ``m_functionCollector.createFunction`` (see elsewhere around the file) overload to deduplicate all that a bit...

Reply:
Had to push another fixup on top, since the function generator you suggested adds newlines before and after the body, so some other tests had to be updated.

Reply:
Just for the record, this could be deduplicated even more by extracting the copying part, which does not depend on any template parameters:

```c++
return Whiskers(R"(
	<copyRoutine>
	<?cleanup>mstore(add(dst, length), 0)</cleanup>
)")
("cleanup", _cleanup)
("<copyRoutine>", ...)
```

Not sure I'd say it's better though. It does make it more fragmented. So just leaving it here as something to consider. I'm fine with the current version too.

Reply:
The version that uses the other ``createFunction`` overload that's in the PR now is good enough I'd say.